classes/citation/Repository.php

```diff
@@ -0,0 +1,288 @@
+     *
+     * @hook Citation::importCitations::before [[$publicationId, $existingCitations, $rawCitations]]
+     * @hook Citation::importCitations::after [[$publicationId, $existingCitations, $importedCitations]]
+     */
+    public function importCitations(int $publicationId, ?string $rawCitationList): void
+    {
+        $existingCitations = $this->getByPublicationId($publicationId);
+
+        Hook::call('Citation::importCitations::before', [$publicationId, $existingCitations, $rawCitationList]);
+
+        // Tokenize raw citations
+        $citationTokenizer = new CitationListTokenizerFilter();
+        $citationStrings = $rawCitationList ? $citationTokenizer->execute($rawCitationList) : [];
+
+        $existingRawCitations = collect($this->getByPublicationId($publicationId))
+            ->map(fn(Citation $citation) => $citation->getData('rawCitation'))
+            ->all();
+
+        if ($existingRawCitations !== $citationStrings) {
+            $importedCitations = [];
+            $this->deleteByPublicationId($publicationId);
+            if (is_array($citationStrings) && !empty($citationStrings)) {
+                foreach ($citationStrings as $seq => $rawCitationString) {
+                    if (!empty(trim($rawCitationString))) {
+                        $citation = new Citation();
+                        $citation->setRawCitation($rawCitationString);
+                        $citation->setData('publicationId', $publicationId);
+                        $citation->setSequence($seq + 1);
+                        $citation->setData('isProcessed', false);
+                        $citationId = $this->dao->insert($citation);
+                        $importedCitations[] = $citation;
+                        $this->addJobForCitation($citationId);
+                    }
+                }
+            }
+
+            Hook::call('Citation::importCitations::after', [$publicationId, $existingCitations, $importedCitations]);
+        }
+    }
+
+    /**
+     * Import citations from a raw citation list of the particular publication to existing citations.
+     */
+    public function importAdditionalCitations(int $publicationId, ?string $rawCitationList): string
+    {
+        $citationTokenizer = new CitationListTokenizerFilter();
+        $citationStrings = $rawCitationList ? $citationTokenizer->execute($rawCitationList) : [];
+        $existingSeq = collect($this->getByPublicationId($publicationId))
+            ->map(fn(Citation $citation) => $citation->getData('seq'))
+            ->all();
+        $lastSeq = end($existingSeq) ? end($existingSeq) + 1 : 1;
+
+        $rejectedCitations = [];
+        if (is_array($citationStrings) && !empty($citationStrings)) {
+            foreach ($citationStrings as $rawCitationString) {
+                if (!empty(trim($rawCitationString))) {
+                    if (!$this->existsRawCitation($publicationId, $rawCitationString)) {
```

Comment:
here again, if the string is the same because we call also cleanCitationString() in setRawCitation()